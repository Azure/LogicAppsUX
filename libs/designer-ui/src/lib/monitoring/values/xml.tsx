import Constants from '../../constants';
import CustomEditor, { EditorLanguage } from '../../editor';
import { RawValue } from './raw';
import type { ValueProps } from './types';
import { isXml } from './utils';

export const XmlValue: React.FC<ValueProps> = (props) => {
  const { displayName, value, visible = true } = props;
  if (!visible) {
    return null;
  }

  let valueAsString: string;
  if (isXml(value)) {
    try {
      valueAsString = encoded_atob(value.$content) || Constants.ZERO_WIDTH_SPACE;
    } catch {
      return <RawValue {...props} />;
    }
  } else {
    valueAsString = String(value) || Constants.ZERO_WIDTH_SPACE;
  }

  const height = 19 * Math.min(valueAsString.split('\n').length, 10);

  return (
    <section className="msla-trace-value-label">
      <label className="msla-trace-value-display-name">{displayName}</label>
      <div className="msla-colorizer-json-body">
        <CustomEditor
          defaultValue={valueAsString}
          height={height}
          language={EditorLanguage.xml}
          lineNumbers="off"
          minimapEnabled={false}
          readOnly={true}
        />
      </div>
    </section>
  );
};

function encoded_atob(value: string): string {
  return decodeURIComponent(escape(removeByteOrderMark(atob(value))));
}

// https://en.wikipedia.org/wiki/Byte_order_mark#UTF-8
function removeByteOrderMark(value: string): string {
  return value.replace(/^\u00ef\u00bb\u00bf/, '');
}
