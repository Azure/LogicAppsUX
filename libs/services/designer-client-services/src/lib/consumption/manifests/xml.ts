import { SettingScope } from '@microsoft/utils-logic-apps';
import type { OperationManifest } from '@microsoft/utils-logic-apps';

const connector: any = {
  id: 'connectionProviders/xmlOperations',
  name: 'xmlOperations',
  properties: {
    description: 'XML Operations',
    displayName: 'XML Operations',
  },
};

const settings: any = {
  secureData: {},
  trackedProperties: {
    scopes: [SettingScope.Action],
  },
};

export const xmlTransformManifest = {
  properties: {
    iconUri: 'https://logicappsv2resources.blob.core.windows.net/icons/xml.svg',
    brandColor: '#804998',
    summary: 'Transform XML',
    description: 'Transform XML using XSLT map.',

    inputs: {
      type: 'object',
      required: ['content', 'mapName', 'xsltParameters'],
      properties: {
        containerId: {
          name: 'containerId',
          title: 'Container Id',
          summary: 'Function container',
          description: 'The function container',
          type: 'string',
          hideInUI: true,
          'x-ms-dynamic-values': {
            operationId: 'xslt_get_containers',
            'value-collection': 'value',
            'value-path': 'id',
            'value-title': 'Name',
          },
        },
        functionId: {
          name: 'functionId',
          title: 'Function Id',
          summary: 'Function',
          description: 'The function to use for transform',
          type: 'string',
          hideInUI: true,
          'x-ms-dynamic-values': {
            operationId: 'xslt_get_functions_in_container',
            parameters: {
              containerId: {
                parameter: 'containerId',
              },
            },
            'value-collection': 'value',
            'value-path': 'id',
            'value-title': 'properties/name',
          },
        },
        content: {
          name: 'content',
          title: 'Content',
          summary: 'Content',
          description: 'The XML content to transform',
          required: true,
          type: 'string',
        },
        mapName: {
          name: 'mapName',
          title: 'Map',
          summary: 'Map',
          description: 'The name of the map to use from the associated integration account',
          required: true,
          type: 'string',
          'x-ms-dynamic-values': {
            operationId: 'xslt_get_maps',
            'value-collection': 'value',
            'value-path': 'name',
            'value-title': 'name',
            parameters: {
              $filter: "maptype ne 'liquid'",
            },
          },
        },
        xsltParameters: {
          name: 'xsltParameters',
          title: 'XSLT Parameters',
          summary: 'Map parameters',
          description: 'The map parameters',
          hideInUI: true,
          schema: {
            type: 'object',
            'x-ms-dynamic-schema': {
              operationId: 'xslt_get_map',
              parameters: {
                mapName: {
                  parameter: 'mapName',
                },
              },
              'value-path': 'properties/parametersSchema',
            },
          },
        },
        transformOptions: {
          name: 'transformOptions',
          title: 'Transform Options',
          summary: 'Transform options',
          description: 'The transform options',
          type: 'string',
          visibility: 'advanced',
          'x-ms-editor': 'dropdown',
          'x-ms-editor-options': {
            multiselect: true,
            valueSeparator: ',',
            titleSeparator: ',',
            items: [
              {
                title: 'Disable the byte order mark.',
                value: 'DisableByteOrderMark',
              },
              {
                title: 'Generate text output.',
                value: 'GenerateTextOutput',
              },
              {
                title: 'Apply XSLT output attributes.',
                value: 'ApplyXsltOutputAttributes',
              },
            ],
          },
        },
      },
    },

    outputs: {
      type: 'object',
      properties: {
        body: {
          title: 'Body',
        },
      },
    },
    outputsSchema: {
      outputPaths: [
        {
          outputLocation: ['properties', 'body'],
          name: 'schema',
          schema: 'Value',
        },
      ],
    },
    isOutputsOptional: false,

    connector,
    settings,
  },
} as OperationManifest;

export const xmlValidationManifest = {
  properties: {
    iconUri: 'https://logicappsv2resources.blob.core.windows.net/icons/xml.svg',
    brandColor: '#804998',
    summary: 'XML Validation',
    description: 'Validate XML using schema.',

    inputs: {
      type: 'object',
      required: ['content'],
      properties: {
        content: {
          title: 'Content',
          summary: 'Message Content',
          description: 'The XML content to validate',
          required: true,
          type: 'string',
        },
        schemaName: {
          title: 'Schema Name',
          summary: 'Schema Name',
          description: 'The name of the XML schema to use from the associated integration account',
          type: 'string',
          visibility: 'important',
          'x-ms-dynamic-values': {
            operationId: 'content_and_schema_operation_get_schemas',
            'value-collection': 'value',
            'value-path': 'name',
            'value-title': 'name',
          },
        },
      },
    },

    outputs: {
      type: 'object',
      properties: {
        transformedXml: {
          title: 'Transformed XML',
          type: 'string',
        },
      },
    },

    connector,
    settings,
  },
} as OperationManifest;
