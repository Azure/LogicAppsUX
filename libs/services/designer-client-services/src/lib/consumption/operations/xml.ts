import { xmlGroup } from './operationgroups';

const api = {
  id: xmlGroup.id,
  name: xmlGroup.name,
  displayName: xmlGroup.properties.displayName,
  iconUri: xmlGroup.properties.iconUri,
  brandColor: xmlGroup.properties.brandColor,
  description: xmlGroup.properties.description,
};

export const xmlTransformOperation = {
  name: 'xmlTransform',
  id: 'xmlTransform',
  type: 'xslt',
  properties: {
    api,
    summary: 'Transform XML',
    description: 'Transform XML using XSLT map.',
    visibility: 'Important',
    operationType: 'Xslt',
    brandColor: api.brandColor,
    iconUri: api.iconUri,
  },
};

export const xmlValidationOperation = {
  name: 'xmlValidation',
  id: 'xmlValidation',
  type: 'XmlValidation',
  properties: {
    api,
    summary: 'XML Validation',
    description: 'Validate XML using schema.',
    visibility: 'Important',
    operationType: 'XmlValidation',
    brandColor: api.brandColor,
    iconUri: api.iconUri,
  },
};
