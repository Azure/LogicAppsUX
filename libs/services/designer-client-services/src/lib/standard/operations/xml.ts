export const xmlTransformOperation = {
  name: 'xmlTransform',
  id: 'xmlTransform',
  type: 'xslt',
  properties: {
    api: {
      id: 'connectionProviders/xmlOperations',
      name: 'xmlOperations',
      displayName: 'XML Operations',
      iconUri: 'https://logicappsv2resources.blob.core.windows.net/icons/xml.svg',
      brandColor: '#804998',
      description: 'XML Operations',
    },
    summary: 'Transform XML',
    description: 'Transform XML using XSLT map.',
    visibility: 'Important',
    operationType: 'Xslt',
    brandColor: '#804998',
    iconUri: 'https://logicappsv2resources.blob.core.windows.net/icons/xml.svg',
    annotation: {
      status: 'Preview',
      family: 'xmlOperations',
    },
  },
};

export const xmlValidationOperation = {
  name: 'xmlValidation',
  id: 'xmlValidation',
  type: 'XmlValidation',
  properties: {
    api: {
      id: 'connectionProviders/xmlOperations',
      name: 'xmlOperations',
      displayName: 'XML Operations',
      iconUri: 'https://logicappsv2resources.blob.core.windows.net/icons/xml.svg',
      brandColor: '#804998',
      description: 'XML Operations',
    },
    summary: 'XML Validation',
    description: 'Validate XML using schema.',
    visibility: 'Important',
    operationType: 'XmlValidation',
    brandColor: '#804998',
    iconUri: 'https://logicappsv2resources.blob.core.windows.net/icons/xml.svg',
    annotation: {
      status: 'Preview',
      family: 'xmlOperations',
    },
  },
};
