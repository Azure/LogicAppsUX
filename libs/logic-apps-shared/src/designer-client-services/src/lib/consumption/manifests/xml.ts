import { xmlGroup, xmlValidationOperation } from '../operations';
import type { OperationManifest } from '@microsoft/utils-logic-apps';
import { SettingScope } from '@microsoft/utils-logic-apps';

const settings: any = {
  secureData: {},
  trackedProperties: {
    scopes: [SettingScope.Action],
  },
};

export const xmlTransformManifest = {
  properties: {
    iconUri: xmlValidationOperation.properties.iconUri,
    brandColor: xmlValidationOperation.properties.brandColor,
    summary: 'Transform XML',
    description: 'Transform XML using XSLT map.',

    inputs: {
      type: 'object',
      required: ['content', 'integrationAccount'],
      properties: {
        content: {
          title: 'Content',
          description: 'The XML content to transform',
          type: 'string',
        },
        integrationAccount: {
          type: 'object',
          properties: {
            map: {
              type: 'object',
              properties: {
                name: {
                  title: 'Map Name',
                  type: 'string',
                  description: 'The name of the map to use from the associated integration account',
                  'x-ms-dynamic-list': {
                    dynamicState: {
                      operationId: 'getMapArtifacts',
                      parameters: {
                        $filter: {
                          value: "maptype ne 'liquid'",
                        },
                      },
                    },
                    parameters: {},
                  },
                },
              },
              required: ['name'],
            },
          },
          required: ['map'],
        },
        xsltParameters: {
          title: 'Map Parameters',
          description: 'The map parameters',
          type: 'object',
          'x-ms-dynamic-properties': {
            dynamicState: {
              extension: { operationId: 'getMapSchema' },
              parameters: {},
              isInput: true,
            },
            parameters: {
              mapName: {
                parameterReference: 'integrationAccount.map.name',
                required: true,
              },
            },
          },
        },
        transformOptions: {
          title: 'Transform Options',
          description: 'The transform options',
          type: 'string',
          visibility: 'advanced',
          'x-ms-editor': 'combobox',
          'x-ms-editor-options': {
            multiSelect: true,
            titleSeparator: ',',
            options: [
              {
                displayName: 'Disable the byte order mark.',
                value: 'DisableByteOrderMark',
              },
              {
                displayName: 'Generate text output.',
                value: 'GenerateTextOutput',
              },
              {
                displayName: 'Apply XSLT output attributes.',
                value: 'ApplyXsltOutputAttributes',
              },
            ],
          },
        },
      },
    },

    outputs: {
      type: 'object',
      properties: {
        body: {
          title: 'Transformed XML',
          type: 'string',
          format: 'binary',
        },
      },
    },
    isOutputsOptional: false,

    connector: xmlGroup,
    settings,
  },
} as OperationManifest;

export const xmlValidationManifest = {
  properties: {
    iconUri: xmlValidationOperation.properties.iconUri,
    brandColor: xmlValidationOperation.properties.brandColor,
    summary: 'XML Validation',
    description: 'Validate XML using schema.',

    inputs: {
      type: 'object',
      required: ['content'],
      properties: {
        content: {
          title: 'Content',
          summary: 'Message Content',
          description: 'The XML content to validate',
          type: 'string',
        },
        integrationAccount: {
          type: 'object',
          properties: {
            schema: {
              type: 'object',
              properties: {
                name: {
                  title: 'Schema Name',
                  type: 'string',
                  'x-ms-visibility': 'important',
                  description: 'The name of the XML schema to use from the associated integration account',
                  'x-ms-dynamic-list': {
                    dynamicState: {
                      operationId: 'getSchemaArtifacts',
                      parameters: {},
                    },
                    parameters: {},
                  },
                },
              },
              required: ['name'],
            },
          },
          required: ['schema'],
        },
      },
    },

    outputs: {
      type: 'object',
      properties: {
        body: {
          title: 'Body',
        },
      },
    },
    isOutputsOptional: false,

    connector: xmlGroup,
    settings,
  },
} as OperationManifest;
