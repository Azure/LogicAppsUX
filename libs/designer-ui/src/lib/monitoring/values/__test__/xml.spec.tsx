import renderer from 'react-test-renderer';
import type { ValueProps } from '../types';
import { XmlValue } from '../xml';
import { describe, vi, beforeEach, afterEach, beforeAll, afterAll, it, test, expect } from 'vitest';
describe('ui/monitoring/values/xml', () => {
  let props: ValueProps;

  beforeEach(() => {
    props = {
      displayName: 'XML',
      value: {
        '$content-type': 'application/xml',
        $content: 'PHhtbD48L3htbD4=',
      },
    };
  });

  it('should render XML', () => {
    const tree = renderer.create(<XmlValue {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render XML without the UTF-8 byte order mark (BOM)', () => {
    props.value = {
      '$content-type': 'application/xml',
      $content: '77u/PHhtbD48L3htbD4=', // <xml></xml> with a prepended UTF-8 BOM
    };
    const tree = renderer.create(<XmlValue {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a zero-width space if value is empty', () => {
    props.value = {
      '$content-type': 'application/xml',
      $content: '',
    };
    const tree = renderer.create(<XmlValue {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render XML with encoded UTF-8 Chinese correctly', () => {
    props.value = {
      '$content-type': 'application/xml',
      $content:
        '77u/PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz48cm9vdD48bmFtZSB1c2VybmFtZT0iSlMx55So5oi35ZCNIj7kvaDlpb0gSm9objwvbmFtZT48bmFtZSB1c2VybmFtZT0iTUkx55So5oi35ZCNIj7kuK3mlofmtYvor5U8L25hbWU+PC9yb290Pg==',
    };
    const tree = renderer.create(<XmlValue {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not render when not visible', () => {
    const tree = renderer.create(<XmlValue {...props} visible={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
