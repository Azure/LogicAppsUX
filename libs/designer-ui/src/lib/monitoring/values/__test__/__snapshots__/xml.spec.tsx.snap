// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ui/monitoring/values/xml > should not render when not visible 1`] = `null`;

exports[`ui/monitoring/values/xml > should render XML 1`] = `
<section
  className="msla-trace-value-label"
>
  <label
    className="msla-trace-value-display-name"
  >
    XML
  </label>
  <div
    className="msla-colorizer-json-body"
  >
    <div
      aria-label="XML"
      aria-readonly={true}
      className="msla-colorizer-wrapper"
      role="textbox"
      tabIndex={0}
    >
      <div
        className="buttons"
      >
        <button
          aria-label="Select all text in 'XML'"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Select all text in 'XML'"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-118 ms-Button-icon icon-112"
              data-icon-name="SelectAll"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-1"",
                }
              }
            >
              
            </i>
          </span>
        </button>
        <button
          aria-label="Copy the value of 'XML' to the clipboard"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Copy the value of 'XML' to the clipboard"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-120 ms-Button-icon icon-112"
              data-icon-name="Copy"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-2"",
                }
              }
            >
              
            </i>
          </span>
        </button>
      </div>
      <pre
        className="prism-code language-json"
        style={
          {
            "backgroundColor": "#ffffff",
            "color": "#000000",
          }
        }
      >
        <div
          className="token-line"
          style={
            {
              "backgroundColor": undefined,
              "color": "#000000",
            }
          }
        >
          <span
            className="token plain"
          >
            &lt;xml&gt;&lt;/xml&gt;
          </span>
        </div>
      </pre>
    </div>
  </div>
</section>
`;

exports[`ui/monitoring/values/xml > should render XML with encoded UTF-8 Chinese correctly 1`] = `
<section
  className="msla-trace-value-label"
>
  <label
    className="msla-trace-value-display-name"
  >
    XML
  </label>
  <div
    className="msla-colorizer-json-body"
  >
    <div
      aria-label="XML"
      aria-readonly={true}
      className="msla-colorizer-wrapper"
      role="textbox"
      tabIndex={0}
    >
      <div
        className="buttons"
      >
        <button
          aria-label="Select all text in 'XML'"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Select all text in 'XML'"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-118 ms-Button-icon icon-112"
              data-icon-name="SelectAll"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-1"",
                }
              }
            >
              
            </i>
          </span>
        </button>
        <button
          aria-label="Copy the value of 'XML' to the clipboard"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Copy the value of 'XML' to the clipboard"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-120 ms-Button-icon icon-112"
              data-icon-name="Copy"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-2"",
                }
              }
            >
              
            </i>
          </span>
        </button>
      </div>
      <pre
        className="prism-code language-json"
        style={
          {
            "backgroundColor": "#ffffff",
            "color": "#000000",
          }
        }
      >
        <div
          className="token-line"
          style={
            {
              "backgroundColor": undefined,
              "color": "#000000",
            }
          }
        >
          <span
            className="token plain"
          >
            &lt;?xml version="1.0" encoding="utf-8"?&gt;&lt;root&gt;&lt;name username="JS1用户名"&gt;你好 John&lt;/name&gt;&lt;name username="MI1用户名"&gt;中文测试&lt;/name&gt;&lt;/root&gt;
          </span>
        </div>
      </pre>
    </div>
  </div>
</section>
`;

exports[`ui/monitoring/values/xml > should render XML without the UTF-8 byte order mark (BOM) 1`] = `
<section
  className="msla-trace-value-label"
>
  <label
    className="msla-trace-value-display-name"
  >
    XML
  </label>
  <div
    className="msla-colorizer-json-body"
  >
    <div
      aria-label="XML"
      aria-readonly={true}
      className="msla-colorizer-wrapper"
      role="textbox"
      tabIndex={0}
    >
      <div
        className="buttons"
      >
        <button
          aria-label="Select all text in 'XML'"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Select all text in 'XML'"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-118 ms-Button-icon icon-112"
              data-icon-name="SelectAll"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-1"",
                }
              }
            >
              
            </i>
          </span>
        </button>
        <button
          aria-label="Copy the value of 'XML' to the clipboard"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Copy the value of 'XML' to the clipboard"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-120 ms-Button-icon icon-112"
              data-icon-name="Copy"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-2"",
                }
              }
            >
              
            </i>
          </span>
        </button>
      </div>
      <pre
        className="prism-code language-json"
        style={
          {
            "backgroundColor": "#ffffff",
            "color": "#000000",
          }
        }
      >
        <div
          className="token-line"
          style={
            {
              "backgroundColor": undefined,
              "color": "#000000",
            }
          }
        >
          <span
            className="token plain"
          >
            &lt;xml&gt;&lt;/xml&gt;
          </span>
        </div>
      </pre>
    </div>
  </div>
</section>
`;

exports[`ui/monitoring/values/xml > should render a zero-width space if value is empty 1`] = `
<section
  className="msla-trace-value-label"
>
  <label
    className="msla-trace-value-display-name"
  >
    XML
  </label>
  <div
    className="msla-colorizer-json-body"
  >
    <div
      aria-label="XML"
      aria-readonly={true}
      className="msla-colorizer-wrapper"
      role="textbox"
      tabIndex={0}
    >
      <div
        className="buttons"
      >
        <button
          aria-label="Select all text in 'XML'"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Select all text in 'XML'"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-118 ms-Button-icon icon-112"
              data-icon-name="SelectAll"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-1"",
                }
              }
            >
              
            </i>
          </span>
        </button>
        <button
          aria-label="Copy the value of 'XML' to the clipboard"
          className="ms-Button ms-Button--icon root-109"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          title="Copy the value of 'XML' to the clipboard"
          type="button"
        >
          <span
            className="ms-Button-flexContainer flexContainer-110"
            data-automationid="splitbuttonprimary"
          >
            <i
              aria-hidden={true}
              className="ms-Icon root-105 css-120 ms-Button-icon icon-112"
              data-icon-name="Copy"
              style={
                {
                  "fontFamily": ""FabricMDL2Icons-2"",
                }
              }
            >
              
            </i>
          </span>
        </button>
      </div>
      <pre
        className="prism-code language-json"
        style={
          {
            "backgroundColor": "#ffffff",
            "color": "#000000",
          }
        }
      >
        <div
          className="token-line"
          style={
            {
              "backgroundColor": undefined,
              "color": "#000000",
            }
          }
        >
          <span
            className="token plain"
          >
            ‌
          </span>
        </div>
      </pre>
    </div>
  </div>
</section>
`;
